Yafc.Model/Model/ProductionTable.cs

```diff
@@ -141,13 +141,17 @@ public bool Search(SearchQuery query) {
         /// <param name="ingredientVariantComparer">The comparer to use when deciding which fluid variants to use.</param>
         /// <param name="selectedFuel">If not <see langword="null"/>, this method will select a crafter or lab that can use this fuel, assuming such an entity exists.
         /// For example, if the selected fuel is coal, the recipe will be configured with a burner assembler/lab if any are available.</param>
-        public void AddRecipe(RecipeOrTechnology recipe, IComparer<Goods> ingredientVariantComparer, Goods? selectedFuel = null) {
+        public void AddRecipe(RecipeOrTechnology recipe, IComparer<Goods> ingredientVariantComparer, Goods? selectedFuel = null, Goods? spentFuel = null) {
             RecipeRow recipeRow = new RecipeRow(this, recipe);
             this.RecordUndo().recipes.Add(recipeRow);
-            EntityCrafter? selectedFuelCrafter = selectedFuel?.fuelFor.OfType<EntityCrafter>().Where(e => e.recipes.Contains(recipe)).AutoSelect(DataUtils.FavoriteCrafter);
-            recipeRow.entity = selectedFuelCrafter ?? recipe.crafters.AutoSelect(DataUtils.FavoriteCrafter);
+            EntityCrafter? selectedFuelCrafter = GetSelectedFuelCrafter(recipe, selectedFuel);
+            EntityCrafter? spentFuelRecipeCrafter = GetSpentFuelRecipeCrafter(recipe, spentFuel);
+
+            recipeRow.entity = selectedFuelCrafter ?? spentFuelRecipeCrafter ?? recipe.crafters.AutoSelect(DataUtils.FavoriteCrafter);
             if (recipeRow.entity != null) {
-                recipeRow.fuel = recipeRow.entity.energy.fuels.FirstOrDefault(e => e == selectedFuel) ?? recipeRow.entity.energy.fuels.AutoSelect(DataUtils.FavoriteFuel);
+                recipeRow.fuel = recipeRow.entity.energy.fuels.FirstOrDefault(e => e == selectedFuel)
+                    ?? (spentFuel is not null ? recipeRow.entity.energy.fuels.FirstOrDefault(e => spentFuel?.miscSources.Contains(e) ?? false) : null)
+                    ?? recipeRow.entity.energy.fuels.AutoSelect(DataUtils.FavoriteFuel);
```

Comment:
These three lines look cool, but they are a horror story to debug and read. I suggest a separate function for this whole `if`, where there is no more than two `?` or `??` per line.